enterprise/server/sociartifactstore/sociartifactstore.go

```diff
@@ -174,65 +202,79 @@ func getTargetImageRef(ctx context.Context, image string, platform *rgpb.Platfor
 	remoteDesc, err := remote.Get(imageRef, remoteOpts...)
 	if err != nil {
 		if t, ok := err.(*transport.Error); ok && t.StatusCode == http.StatusUnauthorized {
-			return ctrname.Digest{}, status.PermissionDeniedErrorf("could not retrieve image manifest: %s", err)
+			return v1.Hash{}, status.PermissionDeniedErrorf("could not retrieve image manifest: %s", err)
 		}
-		return ctrname.Digest{}, status.UnavailableErrorf("could not retrieve manifest from remote: %s", err)
+		return v1.Hash{}, status.UnavailableErrorf("could not retrieve manifest from remote: %s", err)
 	}
 
 	targetImg, err := targetImageFromDescriptor(remoteDesc, platform)
 	if err != nil {
-		return ctrname.Digest{}, err
+		return v1.Hash{}, err
 	}
 
 	// Check whether the supplied credentials are sufficient to access the
 	// remote image.
 	if err := checkAccess(ctx, imageRef.Context(), targetImg, authenticator); err != nil {
-		return ctrname.Digest{}, err
+		return v1.Hash{}, err
 	}
 
-	targetImgDigest, err := targetImg.Digest()
+	manifest, err := targetImg.Manifest()
 	if err != nil {
-		return ctrname.Digest{}, err
+		return v1.Hash{}, err
 	}
-	return imageRef.Context().Digest(targetImgDigest.String()), nil
+	return manifest.Config.Digest, nil
 }
 
 func (s *SociArtifactStore) GetArtifacts(ctx context.Context, req *socipb.GetArtifactsRequest) (*socipb.GetArtifactsResponse, error) {
 	ctx, err := prefix.AttachUserPrefixToContext(ctx, s.env)
 	if err != nil {
 		return nil, err
 	}
-	imageRef, err := getTargetImageRef(ctx, req.Image, req.Platform, req.Credentials)
+	configHash, err := getImageManifestConfig(ctx, req.Image, req.Platform, req.Credentials)
 	if err != nil {
 		return nil, err
 	}
-	imageRefDigest := imageRef.DigestStr()
-	exists, err := s.blobstore.BlobExists(ctx, blobKey(imageRefDigest))
-	if err != nil {
-		return nil, err
-	}
-	if exists {
-		return s.getArtifactsFromCache(ctx, imageRefDigest)
-	}
-	// TODO(iain): add a mutex to prevent multiple parallel calls
-	sociIndexDigest, ztocDigests, err := pullAndIndexImage(ctx, req.Image, imageRef.DigestStr())
-	if err != nil {
-		return nil, err
-	}
-	if err = s.writeArtifactsToCache(ctx, imageRefDigest, sociIndexDigest, ztocDigests); err != nil {
+	configHashHex := configHash.Hex
+
+	// Try to only read-pull-index-write once at a time to prevent hammering
+	// the containter registry with a ton of parallel pull requests, and save
+	// apps a bunch of parallel work.
+	workKey := fmt.Sprintf("soci-artifact-store-image-%s", configHashHex)
+	respBytes, err := s.deduper.Do(ctx, workKey, func() ([]byte, error) {
+		exists, err := s.blobstore.BlobExists(ctx, blobKey(configHashHex))
+		if err != nil {
+			return nil, err
+		}
+		var resp *socipb.GetArtifactsResponse
+		if exists {
+			resp, err = s.getArtifactsFromCache(ctx, configHashHex)
```

Comment:
I think I missed this last time around, but you need to deal with cache eviction. If the artifacts are not in the cache anymore, you should trigger a reindexing.

Reply:
Ahh good call, added a check for the ztocs and restructured the code a bit.

There's still technically a race here wherein one (or more) artifacts could be evicted from the cache after returning from here but before the executor reads them. It's not a long window of time, maybe a few seconds, but it is possible. I don't think it's worth trying to address that now but LMK what you think.